feat: remove flag icons

BrocksiNet opened the pull request:
### 1. Why is this change necessary?

The storefront language selector only displayed hardcoded CSS flags for `de-DE`, `en-GB`, and `en-US` locales, resulting in white rectangles for all other languages (e.g., `de-CH`, `fr-FR`, `it-IT`, `de-AT`). This created a poor user experience for shops using languages beyond these three defaults.

Beyond the technical issue, using country flags for language selection is fundamentally problematic:
- **Political sensitivity**: Flags can cause political ambiguity and even insults (see [Flag icons for languages - Criticism](https://en.wikipedia.org/wiki/Flag_icons_for_languages#Criticism))
- **Accessibility**: Text-based language selectors are more accessible for screen readers and users with visual impairments
- **Best practices**: Modern e-commerce platforms have moved away from flags to text-based language selectors

This change aligns Shopware with modern UX best practices while maintaining extensibility for shops that still want to use flags.

### 2. What does this change do, exactly?

- Deprecates hardcoded CSS language flags in `_flags.scss` (wrapped in `@if not feature('v6.8.0.0')` conditional)
- Updates the language selector template to display language names with territory in the dropdown (e.g., "Deutsch (Deutschland)", "English (United Kingdom)") when the `v6.8.0.0` feature flag is active
- Creates extensible Twig blocks (`layout_header_actions_language_widget_content_inner` and `layout_header_actions_languages_widget_form_items_flag_inner`) allowing plugins/themes to add custom flag implementations if desired
- Follows modern e-commerce best practices by showing text-based language selectors

### 3. Describe each step to reproduce the issue or behaviour.

**Before (Issue):**
1. Set up a Shopware shop with a language other than `de-DE` or `en-GB` (e.g., `de-CH` for Swiss German)
2. Navigate to the storefront
3. Open the language selector in the top bar
4. Observe: White rectangle displayed instead of a flag icon

**After (Fixed):**
1. Enable the `v6.8.0.0` feature flag
2. Navigate to the storefront
3. Open the language selector
4. Observe: Clean text display showing "Deutsch (Schweiz)" with no white rectangles

### 4. Please link to the relevant issues (if any).

- closes #4750
- closes #12667

### 5. Checklist

- [ ] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/shopware/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfilled them

BrocksiNet committed:
refactor: Deprecate language flag styles and update Twig blocks for custom flags

- Marked the _flags.scss file as deprecated, indicating it will be removed in v6.8.0.
- Updated the language-widget.html.twig to utilize new Twig blocks for rendering language flags conditionally based on feature availability.

BrocksiNet committed:
chore: Update issue reference format in changelog for storefront language flags removal

BrocksiNet committed:
fix: Update language selection in storefront tests to handle variations of 'Deutsch'

mitelg commented on src/Storefront/Resources/views/storefront/layout/header/actions/language-widget.html.twig:
```diff
@@ -40,8 +48,17 @@
+                                                        {% if not feature('v6.8.0.0') %}
+                                                                <span aria-hidden="true" class="top-bar-list-icon language-flag country-{{ flagCountry }} language-{{ flagLanguage }}"></span>
+                                                        {% else %}
+                                                                {% if language.translationCode.translated.name and language.translationCode.translated.territory %}
```
we should display the name even if the territory is not set. should be two conditions

BrocksiNet committed:
refactor: Simplify language display logic in Twig and update acceptance tests for language variations

- Consolidated language name rendering in language-widget.html.twig to improve clarity.
- Updated acceptance tests to handle variations in language names, ensuring compatibility with both English and German representations.

BrocksiNet commented on tests/acceptance/tests/SalesChannels/StorefrontLanguages.spec.ts:
```diff
@@ -31,8 +31,10 @@ test('Shop customers should be able to view products in different languages.', {
-        await StorefrontHome.languagesMenuOptions.getByText('Deutsch').click();
-        await ShopCustomer.expects(StorefrontHome.languagesDropdown).toContainText('Deutsch');
+        // if you run it with feature flag v6.8.0 enabled, the test instance is in english
+        // @ToDo: Why is the test instance in english when the feature flag is enabled via .env file? Should also be in german.
```
@frobel do you have any clue about that?

frobel replied:
Never heard about that behaviour. What I could imagine, that there is another container with a different language? Or the feature flag behind is doing something different?

BrocksiNet replied:
Update the test to cover all cases with and without the feature flag ...

mitelg commented on changelog/_unreleased/2025-11-04-remove-hardcoded-language-flags.md:
```diff
@@ -0,0 +1,18 @@
+
+* Changed language selector in `src/Storefront/Resources/views/storefront/layout/header/actions/language-widget.html.twig` to display language name with territory in dropdown (e.g., "Deutsch (Deutschland)", "English (United Kingdom)") when `v6.8.0.0` feature flag is active
+___
+# Next Major Version Changes
```
this will be part of the 6.8 upgrade guide. so you need to write it like the removal of deprecations are already happened. 

to announce the deprecation, you should add a `Upgrade information` section

BrocksiNet committed:
test: Enhance storefront language tests to include German domain URL and improve language selection logic

- Updated the test to use a specific German domain URL for better accuracy in language display.
- Modified the language selection step to ensure it works correctly regardless of feature flag settings, allowing for seamless switching between English and German.

BrocksiNet committed:
chore: Update changelog file

BrocksiNet committed:
Merge branch 'trunk' into feat/remove-flags-icons

BrocksiNet committed:
Merge branch 'trunk' into feat/remove-flags-icons

BrocksiNet committed:
test: Refactor storefront language tests for improved readability and functionality

- Simplified the test structure for better clarity.
- Ensured consistent usage of the German domain URL.
- Enhanced the language selection logic to accommodate different scenarios.

dgrothaus-sw commented:
Just to add some context. The Language Pack contains flags for all languages that are part of the Language Pack. The plugin also adds CSS for all flags, so when the Language Pack is installed, the flags are shown as intended.

With the deprecation of the Language Pack and the issue that flags are not shown for custom languages, the removal of the flags is highly appreciated.

BrocksiNet committed:
Merge branch 'trunk' into feat/remove-flags-icons

BrocksiNet committed:
feat: Update to new changelog process

mitelg commented on RELEASE_INFO-6.7.md:
```diff
@@ -8,13 +8,18 @@
 ## Storefront
 
+### Language selector now displays language with territory
+When the `v6.8.0.0` feature flag is active, the language selector in `src/Storefront/Resources/views/storefront/layout/header/actions/language-widget.html.twig` now displays language names with their territory in the dropdown (e.g., "Deutsch (Deutschland)", "English (United Kingdom)") instead of just the language name.
```
I would not speak of the feature flag, but more like that: "with the next major version..."
"the css flags are deprecated and will be removed with..." is also missing.

mitelg commented on UPGRADE-6.8.md:
```diff
@@ -1,4 +1,15 @@
+
+If your theme or plugin relies on these CSS classes to display country flags in the language selector, the flags will no longer display after upgrading to 6.8.0.0.
+
+New extensible Twig blocks `layout_header_actions_language_widget_content_inner` and `layout_header_actions_languages_widget_form_items_flag_inner` have been added and can be used to implement custom flag solutions.
```
as they will be introduced with 6.7.5.0 already, we should not mentioned that for 6.8 again

mitelg commented on UPGRADE-6.8.md:
```diff
@@ -1,4 +1,15 @@
+
+Hardcoded CSS language flags in `src/Storefront/Resources/app/storefront/src/scss/component/_flags.scss` were removed.
+
+If your theme or plugin relies on these CSS classes to display country flags in the language selector, the flags will no longer display after upgrading to 6.8.0.0.
```
isn't that quite obvious? :thinking:
I would remove that sentence

BrocksiNet committed:
feat: Update release info and upgrade file

BrocksiNet committed:
Merge branch 'trunk' into feat/remove-flags-icons

mitelg commented on RELEASE_INFO-6.7.md:
```diff
@@ -8,13 +8,18 @@
 
 ## Storefront
 
+### Language selector now displays language with territory instead of flags
```
```suggestion
### Language selector will display language with territory instead of flags
```

mitelg commented on RELEASE_INFO-6.7.md:
```diff
@@ -8,13 +8,18 @@
 ## Storefront
 
+### Language selector now displays language with territory instead of flags
+The language selector in the storefront will display the language name with its territory in the dropdown menu (e.g. "Deutsch (Deutschland)", "English (United Kingdom)"). The CSS for flag icons is deprecated and will be removed with Shopware 6.8.0.0.
```
```suggestion
The language selector in the storefront will display the language name with its territory in the dropdown menu (e.g. "Deutsch (Deutschland)", "English (United Kingdom)") with the next major version.
The CSS for flag icons is deprecated and will be removed with Shopware 6.8.0.0.
```

mitelg committed:
fix: Introduce generic ID structure type to reduce ID checks (#13334)

BrocksiNet committed:
fix: Correct language selector description in release notes to clarify upcoming changes

jozsefdamokos committed:
fix: import export profile search by technical name (#13383)

mitelg commented on RELEASE_INFO-6.7.md:
```diff
@@ -8,13 +8,19 @@
+The language selector in the storefront will display the language name with its territory in the dropdown menu (e.g. "Deutsch (Deutschland)", "English (United Kingdom)") with the next major version.
+The CSS for flag icons is deprecated and will be removed with Shopware 6.8.0.0.
+
+New extensible Twig blocks `layout_header_actions_language_widget_content_inner` and `layout_header_actions_languages_widget_form_items_flag_inner` have been added to allow custom flag implementations.
```
```suggestion
### Language selector will display language with territory instead of flags

The language selector in the storefront will display the language name with its territory in the dropdown menu (e.g. "Deutsch (Deutschland)", "English (United Kingdom)") from the next major version on.
The CSS for flag icons is deprecated and will then be removed.

New extensible Twig blocks `layout_header_actions_language_widget_content_inner` and `layout_header_actions_languages_widget_form_items_flag_inner` have been added to allow custom flag implementations.
```

mitelg committed:
Update RELEASE_INFO-6.7.md

gecolay committed:
chore: Improve collection functions type safety (#13350)

Co-authored-by: Michael Telgmann <m.telgmann@shopware.com>
Co-authored-by: Michael Telgmann <mitelg@users.noreply.github.com>

jleifeld committed:
feat: convert to Meteor components in dynamic field renderer (#13185)

FlorianKe committed:
fix(admin): settings storefront icon size (#13400)

BrocksiNet committed:
refactor: replace imitation token with JWT (#12828)

BrocksiNet committed:
Merge branch 'trunk' into feat/remove-flags-icons

mitelg commented on UPGRADE-6.8.md:
```diff
@@ -1,4 +1,11 @@
 # 6.8.0.0
+
+## Introduced in 6.7.5.0
```
the structure will be changed with https://github.com/shopware/shopware/pull/13346

you should consider this already

BrocksiNet committed:
Merge branch 'trunk' into feat/remove-flags-icons

mitelg commented on github-issue-primary-key-validation.md:
```diff
@@ -0,0 +1,182 @@
+# Add validation for primary key consistency between database schema and entity definition
```
what is that?

BrocksiNet replied:
ups ... will be removed

mitelg commented on RELEASE_INFO-6.7.md:
```diff
@@ -57,6 +57,12 @@ Payment tokens are no longer deleted immediately after use. A new scheduled task
 
+### Language selector will display language with territory instead of flags
+
+The language selector in the storefront will display the language name with its territory in the dropdown menu (e.g. "Deutsch (Deutschland)", "English (United Kingdom)") from the next major version on.
```
I just had a small discussion with Max on another PR. 

as this is only documenting a breaking change, it needs to be in the upgrade guide only. the new twig blocks are okay here.

BrocksiNet committed:
fix: remove wrong commited file

BrocksiNet committed:
chore: update release and upgrade notes for language selector and flag removal

marcelbrode commented:
@BrocksiNet Oh damn, finally we got rid of those. Thanks a lot!
